crossc-glsl.cc

```diff
@@ -0,0 +1,31 @@
+#include "SPIRV-Cross/spirv_glsl.hpp"
+#include "crossc-priv.h"
+
+namespace {
+	auto get_glsl(crossc_compiler *comp)
+	{
+		return (spirv_cross::CompilerGLSL*)comp->c.get();
+	}
+}
+
+crossc_compiler *crossc_glsl_create(const uint32_t *words, size_t word_count)
+{
+	crossc_compiler *comp = new crossc_compiler;
+	try {
+		comp->c = std::make_unique<spirv_cross::CompilerGLSL>(words, word_count);
+	} catch (const std::exception &e) {
+		comp->error = std::string { e.what() };
+	}
+	return comp;
+}
+
+void crossc_glsl_set_version(crossc_compiler *comp, int version, crossc_glsl_profile profile)
+{
+	if (!comp->c)
+		return;
+	auto glsl = get_glsl(comp);
```

Comment:
The get_glsl() helper shouldn't be necessary here. crossc-hlsl.cc uses get_hlsl() to hide an ugly cast from CompilerGLSL* to CompilerHLSL*, since CompilerGLSL is the base class of the other compilers. In crossc-glsl.cc, you can use CompilerGLSL directly, so ``glsl`` can be replaced by ``comp->c``.

Reply:
Perhaps using a union to also avoid the ugly cast would be a good idea?
```cpp
struct crossc_compiler {
	union {
		std::unique_ptr<spirv_cross::CompilerGLSL> glsl;
		std::unique_ptr<spirv_cross::CompilerHLSL> hlsl;
	};

	std::string output;
	std::string error;
};
```

Reply:
GCC doesn't compile that as-is, because unique_ptr has a non-trivial destructor. There's no way for the compiler to know which union member to destroy when crossc_compiler is destroyed.

I don't mind just having a unique_ptr\<CompilerGLSL>, despite the ugly cast in crossc-hlsl.cc. CompilerGLSL is the base class, so it makes sense, and it allows some code to be shared between both compiler types (thanks to polymorphism, crossc_set_flip_vert_y() and crossc_compile() in crossc.cc work with both the HLSL and GLSL compilers.)

Reply:
Fair enough. My C++ is very rusty and I'm thinking about this in a way too "C" way. Sorry about that. I'll just keep the HLSL cast and remove the useless helper for GLSL, then.

Reply:
Thanks. (I'm more comfortable with C too, but I figured if I'm forced to use C++, I might as well use smart pointers.)